### ISSUE TITLE
[UX] Add an 

### ISSUE THREAD

**Issue by jenlampton** [backdrop/backdrop-issues#2960]

**Comment by olafgrabienski** [comment]
Another idea (don't know if exportable): single content types including configuration, field and display settings

**Comment by jenlampton** [comment]
I think we have an issue for that already...  somewhere...

**Comment by olafgrabienski** [comment]
Guess I found it: [[CMI] Export/Import an entire content type (or other entity type) with its fields all at once](https://github.com/backdrop/backdrop-issues/issues/408)

Other related issues:

- [[META][DX] CMI: Introduce "export config" links throughout the admin UI that open a modal with the code ready to be copied](https://github.com/backdrop/backdrop-issues/issues/1833)
- [Add "Export" link to drop-button for layouts](https://github.com/backdrop/backdrop-issues/issues/2116) (duplicate of this issue, I'm going to close it)

**Comment by olafgrabienski** [comment]
@jenlampton I've updated the issue description to add links to sub-issues.

I saw also that in (the duplicate issue) #1833 there are some more ideas which we might add to the issue description.

**Comment by jenlampton** [comment]
I think this will be a fantastic UX improvement, so adding the milestone for 1.10. We'll see how many of these sub-issues will be ready by then, but let's give it a go :)

**Comment by jenlampton** [comment]
Didn't make it for 1.10, bumping milestone.

**Comment by klonos** [comment]
This interests me a great deal. Do we have any UI guidelines/mockups/ideas of how this should look and work? I think that the place where this export functionality is located should be consistent throughout the admin UI, similar to how we have the "+Add XYZ" links at the top-left for listing pages.

**Comment by jenlampton** [comment]
I think we should stick with the current pattern of putting the drop-button in with the other actions. For most things it's self explanatory. See views for an example of how the rest of core should look & work.

**Comment by klonos** [comment]
Sure, so Views has an "Export" option in each drop-down for each individual view listed under `/admin/structure/views`, and it also provides an "Export view" option in the drop-down menu that is available within each individual view config page (top-right). I would appreciate an additional "Export all views" option in `/admin/structure/views`.

So similarly, say for the content types listing in `/admin/structure/types` I would expect an "Export" option in each drop-down for each individual content type, and also an "Export all content types" drop-down menu at the top-right. Not sure where to place a respective "Export content type" option within each individual content type "Configure" page though (to match what we have in the Views UI that is).

**Comment by jenlampton** [comment]
I've filed a single PR for this that will include many of the systems that need export links, since they all seemed to be trivial. https://github.com/backdrop/backdrop/pull/2463

**Comment by klonos** [comment]
Tested this on the PR sandbox and it works as expected, so RTBC on that aspect. But...

Minor nit: you have changed all the export button links on the views listing page to use a query, but you have not changed the export buttons when configuring each view individually.

Bigger nit (or a question): why use a query instead of `/admin/config/development/configuration/single/export/[group_name]/[config_name]`?

**Comment by docwilmot** [comment]
This is brilliant. Agree with @klonos questions though.

**Comment by jenlampton** [comment]
> Views has an "Export" option in each drop-down for each individual view listed under /admin/structure/views

This PR fixes that one to use the new "supported" link format. 

> and it also provides an "Export view" option in the drop-down menu that is available within each individual view config page

Oops, missed that one. Will push an update!

> I would appreciate an additional "Export all views" option in /admin/structure/views.

Sure, please open an issue :) 

> I would expect an "Export" option in each drop-down for each individual content type

Yep, this PR adds those.

> and also an "Export all content types" drop-down menu at the top-right.

Please also add this request to the new issue - it sounds like a new concept and we don't yet have the ability to do this (more than one config file, but not all of them) in core.

> Not sure where to place a respective "Export content type" option within each individual content type "Configure" page though

I'm not sure this is necessary...  In fact, I don't think it would be a good idea until we solve https://github.com/backdrop/backdrop-issues/issues/408 because that's what people really want when they export a "Content type" not just the type name and settings, which is what this change will provide.

> Minor nit: you have changed all the export button links on the views listing page to use a query, but you have not changed the export buttons when configuring each view individually.

yes, missed that one :) will update shortly!

> Bigger nit (or a question): why use a query instead of /admin/config/development/configuration/single/export/[group_name]/[config_name]?

Well, 1) Because @quicksketch told me to. But 2) passing in values in the URL would require some major updating to the menu entries to accept the new parameters, and also because our group names have uppercase characters and spaces we'd probably want to add a bunch of transformations, etc. Basically - this was the easiest / least change way to do it.

**Comment by docwilmot** [comment]
> passing in values in the URL would require some major updating to the menu entries to accept the new parameters

but `config_export_single_form(array $form, array &$form_state, $config_group = NULL, $config_name = NULL)` already passes those url parts as parameters. So visiting URL `admin/config/development/configuration/single/export/ANYGROUPNAME/ANY_CONFIG` already works.

TBH I thought it shouldnt, but it does.

**Comment by jenlampton** [comment]
Yes, that's the hack I was using before for Views. It surprized @quicksketch too.

**Comment by jenlampton** [comment]
The issue is that the menu handler doesn't expect those values. This only works because of the weird Drupal/Backdrop bug that if the path `foo` exists, then `foo/bar` will "work" as will "foo/bar/baz" instead of properly 404ing.

**Comment by klonos** [comment]
Makes sense @jenlampton 👍 ...besides, using a query will allow to easily pass multiple things (groups and configs) as parameters if we decide to implement #408 or similar.

**Comment by quicksketch** [comment]
This looks good, and I like the new `$_GET` approach instead of tacking onto the URL as a path.

One thing that concerns me, should we provide a link to export content types when you can't _really_ import it into another site? As exporting a content type doesn't include any fields. There's an existing issue for this at https://github.com/backdrop/backdrop-issues/issues/408.

Should we just go ahead and provide the Export link anyway, even though it probably won't do what people want (yet)?

**Comment by quicksketch** [comment]
Views UI includes a test to check for the export link. That needs to be updated. I don't think my above comment needs to hold this up, but the tests need to pass.

**Comment by jenlampton** [comment]
> Should we just go ahead and provide the [content type] Export link anyway, even though it probably won't do what people want (yet)?

If this is a sticking point I say we include all the other links, and punt on this one. I agree it may cause a UX issue for people where there wasn't one before.

**Comment by klonos** [comment]
> One thing that concerns me, should we provide a link to export content types when you can't really import it into another site? As exporting a content type doesn't include any fields. There's an existing issue for this at #408.

This is also mentioned in #3475, as it seems that users generally expect CMI to work similarly to Features (automatically export dependencies along with the thing selected).
